### PR TITLE
Make MLflow project Databricks spark job supporting setting param-list and entry-point CLI argument

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/12854?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12854/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12854
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

For MLflow project Databricks spark job, allow user to add "entry_point" configuration , user can add parameter and any python command. Example:

https://github.com/WeichenXu123/mlflow-job-test2

```
name: tutorial

name: tutorial

databricks_spark_job:
    python_libraries: # dependencies to be installed as databricks cluster libraries
      - mlflow==2.4.1
      - scikit-learn
      - pandas
      - dynaconf
      - numpy

entry_points:
  main:
    parameters:
      model_name: {type: string, default: model}
      build_id: {type: string, default: local}
      timestamp: {type: string, default: local}
      environment: {type: string, default: local}
      config: {type: string, default: 'local'}
      script_name: {type: string, default: train.py}
    command: "python {script_name} {model_name} {build_id} {timestamp} {environment} {config}"
```

Project run command:

```
cd <PROJECT_DIR>
export MLFLOW_TRACKING_URI=databricks
mlflow run . -b databricks --backend-config cluster-spec.json\
 -P script_name=train.py -P model_name=model123\   # pass param values here
 --experiment-id <experiment-id>
```

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Make MLflow project Databricks spark job supporting setting param-list and entry-point CLI argument

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [X] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/projects/_project_spec.py [hunk 1]
```diff
@@ -42,13 +42,35 @@ def load_project(directory):
...
+        if python_file is None and not entry_points:
+            raise MlflowException(
+                "Databricks Spark job requires either 'databricks_spark_job.python_file' "
+                "setting or 'entry_points' setting."
+            )
+        if python_file is not None and entry_points:
+            raise MlflowException(
+                "Databricks Spark job does not allow 'databricks_spark_job.python_file' "
+                "setting and 'entry_points' setting coexisting."
+            )
```
Comment (BenWilson2) [inline mlflow/projects/_project_spec.py:65]:
```suggestion
                "Databricks Spark job does not allow setting both 'databricks_spark_job.python_file' "
                "setting and 'entry_points'."
            )
```

File: mlflow/projects/_project_spec.py [hunk 2]
```diff
@@ -42,13 +42,35 @@ def load_project(directory):
...
+
+        for entry_point in entry_points.values():
+            for param in entry_point.parameters.values():
+                if param.type == "path":
+                    raise MlflowException(
+                        "Databricks Spark job does not support entry point parameter of 'path' "
+                        f"type but got 'path' type parameter '{param.name}'."
+                    )
```
Comment (BenWilson2) [inline mlflow/projects/_project_spec.py:73]:
```suggestion
                        "Databricks Spark job does not support entry point parameter of 'path' "
                        f"type. '{param.name}' value type is invalid."
                    )
```

File: mlflow/projects/_project_spec.py [hunk 3]
```diff
@@ -202,7 +218,17 @@ def __init__(
...
-            return None
+            if self.databricks_spark_job_spec.python_file is not None:
+                # If Databricks Spark job is configured with python_file field,
+                # it does not need to configure entry_point section
+                # and the 'entry_point' param in 'mlflow run' command is ignored
+                return None
+
+            if self._entry_points is None or entry_point not in self._entry_points:
+                raise MlflowException(
+                    f"The entry point '{entry_point}' is not defined in Databricks spark job "
```
Comment (BenWilson2) [inline mlflow/projects/_project_spec.py:229]:
```suggestion
                    f"The entry point '{entry_point}' is not defined in the Databricks spark job "
```

File: mlflow/projects/databricks.py [hunk 4]
```diff
@@ -281,8 +283,28 @@ def run_databricks_spark_job(
             "=== Running databricks spark job of project %s on Databricks ===", project_uri
         )
 
+        if project_spec.databricks_spark_job_spec.python_file is not None:
+            if entry_point != "main" or parameters:
+                _logger.warning(
+                    "You configured Databricks spark job python_file and parameters in "
+                    "MLProject file databricks_spark_job section, the '--entry-point' "
+                    "and '--param-list' arguments specified in 'mlflow run' command are ignored."
```
Comment (BenWilson2) [inline mlflow/projects/databricks.py:291]:
```suggestion
                    "You configured Databricks spark job python_file and parameters within the "
                    "MLProject file's databricks_spark_job section. '--entry-point' "
                    "and '--param-list' arguments specified in the 'mlflow run' command are ignored."
```

File: mlflow/projects/databricks.py [hunk 5]
```diff
@@ -281,8 +283,28 @@ def run_databricks_spark_job(
...
+                )
+            job_code_file = project_spec.databricks_spark_job_spec.python_file
+            job_parameters = project_spec.databricks_spark_job_spec.parameters
+        else:
+            command = project_spec.get_entry_point(entry_point).compute_command(parameters, None)
+            command_splits = command.split(" ")
+            if command_splits[0] != "python":
+                raise MlflowException(
+                    "Databricks spark job only supports 'python' command in entry point "
```
Comment (BenWilson2) [inline mlflow/projects/databricks.py:300]:
```suggestion
                    "Databricks spark job only supports 'python' command in the entry point "
```

**Comment by BenWilson2** [comment]
Directionally, this looks good. 
Let's add instructions within the docs to illustrate this additional functionality and how conflicting configurations are handled (ensuring that exactly one of `entry_points` or `python_file` are provided, the inability to specify a path based entry point, and how entry point configurations are ignored if the `python_file` configuration is supplied) here: https://docs.databricks.com/en/mlflow/projects.html#mlflow-databricks-spark-job-project-format

**Comment by WeichenXu123** [comment]
@BenWilson2 Doc PR: https://github.com/databricks-eng/docs/pull/17412

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**